plugins/fileservices/fileservices.cpp

```diff
@@ -2726,70 +2726,152 @@ FILESERVICES_API char *  FILESERVICES_CALL fsfResolveHostName(const char *hostna
+}
+
+static void implementMoveExternalFile(ICodeContext *ctx,const char *location,
+    const char *frompath,const char *topath,const char *planename)
+{
     StringBuffer s("MoveExternalFile ('");
-    s.append(location).append(',').append(frompath).append(',').append(topath).append(") done");
-    WUmessage(ctx,SeverityInformation,NULL,s.str());
-    AuditMessage(ctx,"MoveExternalFile",frompath,topath);
+    try
+    {
+        Owned<IPropertyTree> plane = checkPlaneOrHost(planename,location);
+        RemoteFilename fromrfn,torfn;
+        checkExternalFilePath(ctx,plane,location,frompath,true,true,fromrfn);
+        checkExternalFilePath(ctx,plane,location,topath,false,true,torfn);
+
+        Owned<IFile> fileto = createIFile(torfn);
+        if (fileto->exists())
+            throw makeStringExceptionV(-1,"Destination %s already exists",topath);
+        fileto.clear();
+        Owned<IFile> file = createIFile(fromrfn);
+        file->move(topath);
+        s.append(location).append(',').append(frompath).append(',').append(topath).append(") done");
+        AuditMessage(ctx,"MoveExternalFile",frompath,topath);
+    }
+    catch (IException *e)
+    {
+        s.append(location).append(',').append(frompath).append(',').append(topath).append("), ");
+        e->errorMessage(s);
+        e->Release();
+    }
+    WUmessage(ctx,SeverityInformation,nullptr,s.str());
+}
+
+FILESERVICES_API void  FILESERVICES_CALL fsMoveExternalFile(ICodeContext *ctx,const char *location,const char *frompath,const char *topath)
+{
+    implementMoveExternalFile(ctx,location,frompath,topath,nullptr);
+}
+
+FILESERVICES_API void  FILESERVICES_CALL fsMoveExternalFile_v2(ICodeContext *ctx,const char *location,
+    const char *frompath,const char *topath,const char *planename)
+{
+    implementMoveExternalFile(ctx,location,frompath,topath,planename);
 }
 
 FILESERVICES_API void  FILESERVICES_CALL fsDeleteExternalFile(ICodeContext * ctx,const char *location,const char *path)
 {
-    SocketEndpoint ep(location);
-    if (ep.isNull())
-        throw MakeStringException(-1,"fsDeleteExternalFile: Cannot resolve location %s",location);
-    CDfsLogicalFileName lfn;
-    lfn.setExternal(location,path);
-    checkExternalFileRights(ctx,lfn,false,true);
-    RemoteFilename rfn;
-    rfn.setPath(ep,path);
-    Owned<IFile> file = createIFile(rfn);
-    file->remove();
     StringBuffer s("DeleteExternalFile ('");
-    s.append(location).append(',').append(path).append(") done");
-    WUmessage(ctx,SeverityInformation,NULL,s.str());
-    AuditMessage(ctx,"DeleteExternalFile",path);
+    try
+    {
+        SocketEndpoint ep(location);
+        if (ep.isNull())
+            throw makeStringExceptionV(-1,"Cannot resolve location %s",location);
+        CDfsLogicalFileName lfn;
+        lfn.setExternal(location,path);
+
+        RemoteFilename rfn;
+        lfn.getExternalFilename(rfn);
+        Owned<IFile> file = createIFile(rfn);
+        if (!file->exists())
```

Comment:
strictly, I think this is a security leak. It means unauthorized users can discover which files do/don't exist.

It should not leak this information for unauthorized users.
To overcome this, I think you need to assume it's a file and validate the parent scope 1st.
Then check if directory, and only then check the tail if it is.